lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -406,9 +407,227 @@ struct ConvolutionToConv2dPattern
+    std::vector<int64_t> spatial_dims;
+    for (int64_t i = 0;
+         i < static_cast<int64_t>(op.getWindowDimensions().size()); i++) {
+      if (op.getWindowDimensions()[i] > 1) {
+        spatial_dims.push_back(i);
+      }
+    }
+    return spatial_dims;
+  }
+
+  LogicalResult canDecompose2DPoolingOp(ttir::PoolingOp op) const {
+
+    // Window dimensions must be 4 in length
+    if (op.getWindowDimensions().size() != 4) {
+      return failure();
+    }
+
+    // Window strides must be 4 in length
+    if (op.getWindowStrides().size() != 4) {
+      return failure();
+    }
+
+    // Operand rank(s) must be 4
+    for (Value operand : op.getInputs()) {
+      auto operand_type = mlir::cast<mlir::RankedTensorType>(operand.getType());
+      if (operand_type.getRank() != 4) {
+        return failure();
+      }
+    }
+
+    // Exactly two of the window dimensions must be greater than 1
+    std::vector<int64_t> true_window_dimensions_indices = getSpatialDims(op);
+
+    if (true_window_dimensions_indices.size() != 2) {
+      return failure();
+    }
+
+    // Exactly two of the window strides must be greater than 1
+    std::vector<int64_t> true_window_stride_indices;
+    for (int64_t i = 0; i < static_cast<int64_t>(op.getWindowStrides().size());
+         i++) {
+      if (op.getWindowStrides()[i] > 1) {
+        true_window_stride_indices.push_back(i);
+      }
+    }
+
+    if (true_window_stride_indices.size() != 2) {
+      return failure();
+    }
+
+    // The indices of the true window dimensions and strides must be the same
+    if ((true_window_dimensions_indices[0] != true_window_stride_indices[0] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[1]) &&
+        (true_window_dimensions_indices[0] != true_window_stride_indices[1] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[0])) {
+      return failure();
+    }
+
+    // Padding must be 8 in length
+    if (op.getPadding().size() != 8) {
+      return failure();
+    }
+
+    return success();
+  }
+
+  template <typename PoolOpType>
+  void rewritePool2d(ttir::PoolingOp op, OpAdaptor adaptor,
+                     ConversionPatternRewriter &rewriter) const {
+
+    const int64_t SPATIAL_H = -3;
+    const int64_t SPATIAL_W = -2;
+    const int64_t NON_SPATIAL = -1;
+
+    auto inputType =
+        mlir::cast<RankedTensorType>(adaptor.getInputs()[0].getType());
+    assert(inputType.getRank() == 4 && "Input must be 4D tensor");
+    std::vector<int64_t> desired_layout(inputType.getRank(), NON_SPATIAL);
+    desired_layout[inputType.getRank() - 3] = SPATIAL_H;
+    desired_layout[inputType.getRank() - 2] = SPATIAL_W;
+
+    int64_t non_spatial_count = 0;
+    for (int64_t i = 0; i < static_cast<int64_t>(desired_layout.size()); i++) {
+      if (desired_layout[i] == NON_SPATIAL) {
+        desired_layout[i] = non_spatial_count;
+        non_spatial_count++;
+      }
+    }
```

Comment:
Do we need to calculate this? If rank is 4, and we manually set 2 dims to spatial, this is always going to be 2?

Reply:
We calculate this again a few lines below, but I don't think it's used anywhere?

Reply:
I'll remove this, its leftover from when I was trying out a different method of doing this.

Reply:
Actually my mistake. I do need these. I want to assign the non-spatial dims different values in the same order from left to right so that we can accurately generate the correct transposes to make the current layout match the desired layout.